data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/PostStateValidators.java

```diff
@@ -0,0 +1,136 @@
+              RequestBody::setIds)
+          .withOptionalField(
+              "statuses",
+              DeserializableTypeDefinition.listOf(STRING_TYPE),
+              RequestBody::getMaybeStringStatuses,
+              RequestBody::setStatuses)
+          .build();
+
+  private final ChainDataProvider chainDataProvider;
+
+  public PostStateValidators(final DataProvider dataProvider) {
+    this(dataProvider.getChainDataProvider());
+  }
+
+  PostStateValidators(final ChainDataProvider provider) {
+    super(
+        EndpointMetadata.post(GetStateValidators.ROUTE)
+            .operationId("postStateValidators")
+            .summary("Get validators from state")
+            .description(
+                "Returns filterable list of validators with their balance, status and index.")
+            .pathParam(PARAMETER_STATE_ID)
+            .requestBodyType(REQUEST_TYPE)
+            .tags(TAG_BEACON)
+            .response(SC_OK, "Request successful", GetStateValidators.RESPONSE_TYPE)
+            .withNotFoundResponse()
+            .build());
+    this.chainDataProvider = provider;
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final Optional<RequestBody> requestBody = request.getOptionalRequestBody();
+    final List<String> validators = requestBody.map(RequestBody::getIds).orElse(List.of());
+    final List<StatusParameter> statusParameters =
+        requestBody.map(RequestBody::getStatuses).orElse(List.of());
+
+    final Set<ValidatorStatus> statusFilter = getApplicableValidatorStatuses(statusParameters);
+
+    SafeFuture<Optional<ObjectAndMetaData<List<StateValidatorData>>>> future =
+        chainDataProvider.getStateValidators(
+            request.getPathParameter(PARAMETER_STATE_ID), validators, statusFilter);
+
+    request.respondAsync(
+        future.thenApply(
+            maybeData ->
+                maybeData
+                    .map(AsyncApiResponse::respondOk)
+                    .orElseGet(AsyncApiResponse::respondNotFound)));
+  }
+
+  static class RequestBody {
+    private List<String> ids = List.of();
+    private List<StatusParameter> statuses = List.of();
+
+    RequestBody() {}
+
+    public RequestBody(final List<String> ids, final List<StatusParameter> statuses) {
+      this.ids = ids;
+      this.statuses = statuses;
+    }
+
+    public List<String> getIds() {
+      return ids;
+    }
+
+    public Optional<List<String>> getMaybeIds() {
+      return ids.isEmpty() ? Optional.empty() : Optional.of(ids);
+    }
+
+    public void setIds(final Optional<List<String>> ids) {
+      ids.ifPresent(i -> this.ids = i);
+    }
+
+    public List<StatusParameter> getStatuses() {
+      return statuses;
+    }
+
+    public Optional<List<String>> getMaybeStringStatuses() {
+      return statuses.isEmpty()
+          ? Optional.empty()
+          : Optional.of(statuses.stream().map(Enum::name).collect(Collectors.toList()));
```

Comment:
i believe we can just call `toList()` now outside of the collector, which is more efficient...